### ISSUE TITLE
Question: How far shall we enforce "supporting WMF 4.0" for DSC resources?

### ISSUE THREAD

**Issue by johlju** [PowerShell/DscResources#237]
I have not read this anywhere, but might have missed it.

My question is primarily for xSQLServer, but this might be interesting for other resources as well. 
We have a lot of technical debt over at xSQLServer. Keeping support both for WMF 4.0 and WMF 5.x will add more code and test coverage when we now are starting to resolve that. 
Moving to support only WMF 5.x and beyond would of course make life easier for the contributors. But we might be leaving users behind who still aren't allowed to use WMF 5.x (for different reasons).

So... Must we code resource to support WMF 4.0 and up (as far as it can be done, issues might prevent it)?
Or can we start to leave WMF 4.0 behind and new contributions are allowed to leave out support for WMF 4.0 and, more importantly, remove existing WMF 4.0 code?

**Comment by johlju** [comment]
Many resources in xSQLServer expect you to use `PsDscRunAsCredential` to set the credentials to use to login to the SQL Server instance, there is not another credential parameter. If you do not use `PsDscRunAsCredential`, then you need to give SYSTEM account the role System Administrator.

Currently the xSQLServerSetup does not work with the `PsDscRunAsCredential` parameter (issues around that being discussed) so it is used in a way that it is installed using SYSTEM (this is due to a bug). Doing so gives the SYSTEM account automatically System Administrator privileges. In a normal installation, it _does not_ have close to that much rights. Giving SYSTEM that power should only be done if the user expressively set it (which you normally never do).
But since this bug in xSQLServerSetup exists, and the resource is installed with SYSTEM, then all resources works without the `PsDscRunAsCredential`.

When we fix the bug in xSQLServerSetup, all other resources need Credential parameter to be able to work with WMF 4.0. Or we again need to give SYSTEM account too high of rights. But if we choose to do so the resource that should set the permission, xSQLServerPermission, does only support `PsDscRunAsCredential`. So catch 22.

So to support WMF 4.0 and WMF 5.0, and keep the permission at the right level, we need to fix a Credential parameter in each resource, and as well make sure it still can use `PsDscRunAsCredential`. And for those resource that does not work with `PsDscRunAsCredential` parameter we need to add that support.
It would be less work to make sure that just the parameter `PsDscRunAsCredential` is supported in each resource.

On top of that, if we get a class based resources then that will make WMF 4.0 obsolete.
Now there is SQL Server on Linux that I guess we need to support soon. That makes WMF 4.0 obsolete as well.

To be honest, if we are going to support WMF 4.0 fully, with the current lead time, I would say we won't be there until a couple of years. During that time frame the users will have moved to WMF 5.0 (hopefully), and probably have PowerShell xNext1 and vNext2 been released, and then maybe we got even new functionality that start to make WMF 5.x obsolete 😉 

There are _a lot_ of issues in the back log and _a lot_ of new functionality that must be added to the resource as well. We have just scratched the surface of the functionality that need to be created.
As a maintainer I would rather focus on that, giving users more functionality, than adding WMF 4.0 support. But at the same time I don't like the idea of leaving users behind (thus this issue).

But with all that said. If this discussion concludes in that we must support WMF 4.0. Then we focus on that.

**Comment by PlagueHO** [comment]
@johlju - ah I see! I completely understand. I can see why supporting both models would require a whole lot of extra codes and tests.

The xDFS resource module would have the same problem - it requires PsDscRunAsCredential. But I chose to never even implement WMF4.0 support (I don't actually know if I could have made it support WMF 4.0 anyway).

Could resource modules where this sort of problem have a milestone defined - say the next major version (v5.0.0.0 in the case of xSQLServer) - have WMF4.0 dropped. Anyone who is still on WMF 4.0 can still use the v4.0.0.0 version but if they want new features or fixes beyond that point they should be upgrading.

Perhaps we could put together a list of resource modules that are having this issue and then determine if it is worth dropping WMF4.0 support in a future version release - and specify which release. Perhaps the list could be published in this repo and then community feedback requested.

But this does sound like a good topic to be discussed on the next Resource Kit call - @kwirkykat?

**Comment by johlju** [comment]
Closing this since we now have removed support for WMF 4.0 in xSQLServer a few releases back. If there is need for WMF 4.0 an older version must be used (without new functionality and bug fixes). Basically there is to much technical debt and work that needs to be done to keep a fully working WMF 4.0 support. I'd rather look ahead than back.

**Comment by BrianFarnhill** [comment]
So I don't think supporting both models needs a whole lot of extra work at all - in fact, we have been doing it in SharePointDsc since day one. The approach is pretty simple - firstly go and have a look at our "[Invoke-SPDscCommand](https://github.com/PowerShell/SharePointDsc/blob/dev/Modules/SharePointDsc/Modules/SharePointDsc.Util/SharePointDsc.Util.psm1#L258)" cmdlet - this is the brains of the whole operation. Basically, it will detect if the code is running as the local system account, or as a specific account (which means PsDscCredential was called). If it's running as the system account we use a supplied credential and use a CredSSP local session to run code as the specified user. 

What this means is that we have a parameter called "InstallAccount" on each of our resources, which is the WMF 4 compatible version of how to do this. You can see it in action in most of our resources, but have a look at [SPFeature](https://github.com/PowerShell/SharePointDsc/blob/dev/Modules/SharePointDsc/DSCResources/MSFT_SPFeature/MSFT_SPFeature.psm1#L36) for a simple look. Basically, we pass the $InstallAccount parameter in to the invoke call every time, but in all WMF5 scenarios this will be null and we'll just execute the code block in place and it's all fine - but if someone uses InstallAccount they will get that local CredSSP session and it'll work just the same.

The drawbacks to this approach - logging goes out the window when you do the CredSSP approach. None of the Write-Verbose calls in the code blocks work if you do it that way (but when we run with PsDscRunAsCredential it works fine as we just invoke the block directly). The second is the performance overhead - I did some initial testing when PS5 released to the new approach and we saw a 30-50% performance increase for not doing the CredSSP sessions, so configs ran a lot faster. 

The pluses though - this is insanely simple to support and implement in our resources, just wrap the code you need to ensure runs as a specific user in the Invoke command we have there and you're good to go. From a testing perspective things get pretty easy as well. We wrote some [tests to validate the behaviour of the Invoke command](https://github.com/PowerShell/SharePointDsc/blob/dev/Tests/Unit/SharePointDsc/SharePointDsc.Util.Tests.ps1#L31) so that we knew that worked in isolation, then for each individual resource test we just need to mock that Invoke-SPDscCommand to just run the code without the CredSSP bit (which we do in a [helper module](https://github.com/PowerShell/SharePointDsc/blob/dev/Tests/Unit/UnitTestHelper.psm1#L77) that we include in all of our tests so there is no additional work for devs to do when writing tests).

Anyway, long story short - we've found that supporting both versions was honestly not that difficult, and I think you could probably take some of the approaches we have here to simplify doing the same for xSqlServer as well.
